src/device-state/current-state.ts

```diff
@@ -104,170 +109,149 @@ const report = async ({ stateDiff, conf }: StateReport) => {
-};
-
-function reportCurrentState(attempts = 0) {
-	(async () => {
-		const {
-			hardwareMetrics,
-			appUpdatePollInterval: maxDelay,
-			...conf
-		} = await config.getMany([
-			'deviceApiKey',
-			'apiTimeout',
-			'apiEndpoint',
-			'uuid',
-			'localMode',
-			'appUpdatePollInterval',
-			'hardwareMetrics',
-		]);
-
-		reportPending = true;
-		const currentDeviceState = await deviceState.getStatus();
-		// If hardwareMetrics is false, send null patch for system metrics to cloud API
-		const info = {
-			...(hardwareMetrics
-				? await sysInfo.getSystemMetrics()
-				: {
-						cpu_usage: null,
-						memory_usage: null,
-						memory_total: null,
-						storage_usage: null,
-						storage_total: null,
-						storage_block_device: null,
-						cpu_temp: null,
-						cpu_id: null,
-				  }),
-			...(await sysInfo.getSystemChecks()),
-		};
-
-		stateForReport.local = {
-			...stateForReport.local,
+async function generateReport(): Promise<DeviceStatus> {
+	// Get current state
+	const currentDeviceState = await deviceState.getStatus();
+	// Get device metrics and system checks
+	const info = {
+		...((await config.get('hardwareMetrics'))
+			? await sysInfo.getSystemMetrics()
+			: {
+					cpu_usage: null,
+					memory_usage: null,
+					memory_total: null,
+					storage_usage: null,
+					storage_total: null,
+					storage_block_device: null,
+					cpu_temp: null,
+					cpu_id: null,
+			  }),
+		...(await sysInfo.getSystemChecks()),
+	};
+	// Return DeviceStatus object
+	return {
+		local: {
 			...currentDeviceState.local,
 			...info,
-		};
-		stateForReport.dependent = {
-			...stateForReport.dependent,
+		},
+		dependent: {
 			...currentDeviceState.dependent,
-		};
+		},
+	};
+}
 
-		const stateDiff = getStateDiff();
+/**
+ * Start reporting current state
+ *
+ * reportHandler will initialize a loop that trys to patch the current
+ * state using statePatchInterval as the delay between attempts.
+ * If there is an issue with the network request the loop will exponentially
+ * back off until successful then reset interval to original time.
+ */
+export async function reportHandler(attempts = 0): Promise<void> {
+	// Generate latest report repesenting the device's current state
+	const stateForReport = await generateReport();
+	// Calculate differences between current state and last reported state
+	const stateDiff = getStateDiff(lastReportedState, stateForReport);
+	// Try to send report now
+	try {
+		await report(stateDiff);
+	} catch (e) {
+		// Use appUpdatePollInterval as maxDelay since we encountered an error and will backoff now
+		const maxDelay = await config.get('appUpdatePollInterval');
+		// Check if it was an HTTP status error
+		if (e instanceof StatusError) {
+			// Supervisor got an HTTP error from API
+			log.error(
+				`Device state report failure! Status code: ${e.statusCode} - message:`,
+				e.message,
+			);
+			// Use retryAfter as maxDelay if the header is present
+			await backoffPromise(
+				reportHandler,
```

Comment:
I think you might be missing a `return` statement here. If you don't then the report will be called at least twice on the iteration. This also creates a branching point where a parallel recursion is started leading to a memory leak.